In [46]:
import numpy as np
import pandas as pd

In [47]:
df = pd.read_csv("./Airbnb_Open_Data.csv")

/var/folders/9c/9gm285k54tz9ysvymk9gtd400000gn/T/ipykernel_69508/3613748228.py:1: DtypeWarning: Columns (25) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("./Airbnb_Open_Data.csv")


In [48]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 102599 entries, 0 to 102598
Data columns (total 26 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   id                              102599 non-null  int64  
 1   NAME                            102349 non-null  object 
 2   host id                         102599 non-null  int64  
 3   host_identity_verified          102310 non-null  object 
 4   host name                       102193 non-null  object 
 5   neighbourhood group             102570 non-null  object 
 6   neighbourhood                   102583 non-null  object 
 7   lat                             102591 non-null  float64
 8   long                            102591 non-null  float64
 9   country                         102067 non-null  object 
 10  country code                    102468 non-null  object 
 11  instant_bookable                102494 non-null  object 
 12  cancellation_pol

In [38]:
df.isnull().sum()

id                                     0
NAME                                 250
host id                                0
host_identity_verified               289
host name                            406
neighbourhood group                   29
neighbourhood                         16
lat                                    8
long                                   8
country                              532
country code                         131
instant_bookable                     105
cancellation_policy                   76
room type                              0
Construction year                    214
price                                247
service fee                          273
minimum nights                       409
number of reviews                    183
last review                        15893
reviews per month                  15879
review rate number                   326
calculated host listings count       319
availability 365                     448
house_rules     

In [49]:

pd.set_option('display.max_columns', None)


In [50]:
df.head(10)

id                                              NAME      host id  \
0  1001254                Clean & quiet apt home by the park  80014485718   
1  1002102                             Skylit Midtown Castle  52335172823   
2  1002403               THE VILLAGE OF HARLEM....NEW YORK !  78829239556   
3  1002755                                               NaN  85098326012   
4  1003689  Entire Apt: Spacious Studio/Loft by central park  92037596077   
5  1004098         Large Cozy 1 BR Apartment In Midtown East  45498551794   
6  1004650                                   BlissArtsSpace!  61300605564   
7  1005202                                   BlissArtsSpace!  90821839709   
8  1005754                   Large Furnished Room Near B'way  79384379533   
9  1006307                Cozy Clean Guest Room - Family Apt  75527839483   

  host_identity_verified host name neighbourhood group       neighbourhood  \
0            unconfirmed  Madaline            Brooklyn          Kensington   
1               verified     Jenna           Manhattan             Midtown   
2                    NaN     Elise           Manhattan              Harlem   
3            unconfirmed     Garry            Brooklyn        Clinton Hill   
4               verified    Lyndon           Manhattan         East Harlem   
5               verified  Michelle           Manhattan         Murray Hill   
6                    NaN   Alberta            Brooklyn  Bedford-Stuyvesant   
7            unconfirmed      Emma            Brooklyn  Bedford-Stuyvesant   
8               verified    Evelyn           Manhattan      Hell's Kitchen   
9            unconfirmed      Carl           Manhattan     Upper West Side   

        lat      long        country country code instant_bookable  \
0  40.64749 -73.97237  United States           US            False   
1  40.75362 -73.98377  United States           US            False   
2  40.80902 -73.94190  United States           US             True   
3  40.68514 -73.95976  United States           US             True   
4  40.79851 -73.94399  United States           US            False   
5  40.74767 -73.97500  United States           US             True   
6  40.68688 -73.95596  United States           US            False   
7  40.68688 -73.95596  United States           US            False   
8  40.76489 -73.98493  United States           US             True   
9  40.80178 -73.96723  United States           US            False   

  cancellation_policy        room type  Construction year    price  \
0              strict     Private room             2020.0    $966    
1            moderate  Entire home/apt             2007.0    $142    
2            flexible     Private room             2005.0    $620    
3            moderate  Entire home/apt             2005.0    $368    
4            moderate  Entire home/apt             2009.0    $204    
5            flexible  Entire home/apt             2013.0    $577    
6            moderate     Private room             2015.0     $71    
7            moderate     Private room             2009.0  $1,060    
8              strict     Private room             2005.0  $1,018    
9              strict     Private room             2015.0    $291    

  service fee  minimum nights  number of reviews last review  \
0       $193             10.0                9.0  10/19/2021   
1        $28             30.0               45.0   5/21/2022   
2       $124              3.0                0.0         NaN   
3        $74             30.0              270.0    7/5/2019   
4        $41             10.0                9.0  11/19/2018   
5       $115              3.0               74.0   6/22/2019   
6        $14             45.0               49.0   10/5/2017   
7       $212             45.0               49.0   10/5/2017   
8       $204              2.0              430.0   6/24/2019   
9        $58              2.0              118.0   7/21/2017   

   reviews per month  review rate number  calculated h

In [51]:
#Dropping License Column: 

df.drop(columns=['license'], inplace=True)


In [53]:
#Dealing with null value:

# Filing  text-based nulls
df['NAME'] = df['NAME'].fillna('N/A')
df['host_identity_verified'] = df['host_identity_verified'].fillna('unverified')
df['neighbourhood group'] = df['neighbourhood group'].fillna('Unknown')
df['house_rules'] = df['house_rules'].fillna('None')


# Filling boolean or category-based
#df['instant_bookable'] = df['instant_bookable'].fillna('False')
df['cancellation_policy'] = df['cancellation_policy'].fillna('unknown')

# Numeric fixes
df['reviews per month'] = df['reviews per month'].fillna(0)
df['number of reviews'] = df['number of reviews'].fillna(0)




In [55]:
df['host name'] = df['host name'].fillna('N/A')

In [56]:
df['neighbourhood'] = df['neighbourhood'].fillna('Unknown')

In [31]:
df.isnull().sum()

id                                    0
NAME                                  0
host id                               0
host_identity_verified                0
host name                           406
neighbourhood group                   0
neighbourhood                        16
lat                                   8
long                                  8
country                             532
country code                        131
instant_bookable                      0
cancellation_policy                   0
room type                             0
Construction year                   214
price                               247
service fee                         273
minimum nights                      409
number of reviews                     0
last review                       15893
reviews per month                     0
review rate number                  326
calculated host listings count      319
availability 365                    448
house_rules                           0


In [57]:
df['instant_bookable'] = df['instant_bookable'].fillna('False')

In [32]:
# For price/service fee/minimum nights
df['price'] = pd.to_numeric(df['price'], errors='coerce')


In [33]:
df['price'] = df['price'].fillna(df['price'].median())

/Users/saiswethalakkoju/anaconda3/lib/python3.11/site-packages/numpy/lib/_nanfunctions_impl.py:1215: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)


In [34]:
df.isnull().sum()

id                                     0
NAME                                   0
host id                                0
host_identity_verified                 0
host name                            406
neighbourhood group                    0
neighbourhood                         16
lat                                    8
long                                   8
country                              532
country code                         131
instant_bookable                       0
cancellation_policy                    0
room type                              0
Construction year                    214
price                             102599
service fee                          273
minimum nights                       409
number of reviews                      0
last review                        15893
reviews per month                      0
review rate number                   326
calculated host listings count       319
availability 365                     448
house_rules     

In [59]:
# Convert 't'/'f' and any string 'False' to real booleans
df['instant_bookable'] = df['instant_bookable'].replace({'t': True, 'f': False, 'False': False, 'True': True})
df['instant_bookable'] = df['instant_bookable'].astype(bool)


/var/folders/9c/9gm285k54tz9ysvymk9gtd400000gn/T/ipykernel_69508/3329949727.py:2: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['instant_bookable'] = df['instant_bookable'].replace({'t': True, 'f': False, 'False': False, 'True': True})


In [60]:
df['instant_bookable'].value_counts()

instant_bookable
False    51579
True     51020
Name: count, dtype: int64

In [61]:
df['number of reviews'].value_counts()

number of reviews
0.0      15917
1.0      10408
2.0       7175
3.0       5375
4.0       4151
         ...  
567.0        1
592.0        1
797.0        1
966.0        1
300.0        1
Name: count, Length: 476, dtype: int64

In [62]:
df.isnull().sum()

id                                    0
NAME                                  0
host id                               0
host_identity_verified                0
host name                             0
neighbourhood group                   0
neighbourhood                         0
lat                                   8
long                                  8
country                             532
country code                        131
instant_bookable                      0
cancellation_policy                   0
room type                             0
Construction year                   214
price                               247
service fee                         273
minimum nights                      409
number of reviews                     0
last review                       15893
reviews per month                     0
review rate number                  326
calculated host listings count      319
availability 365                    448
house_rules                           0


In [64]:
df_ab = df[df['number of reviews'].notnull()]


In [66]:
group_a = df_ab[df_ab['instant_bookable'] == False]['number of reviews']
group_b = df_ab[df_ab['instant_bookable'] == True]['number of reviews']


In [67]:
mean_a = group_a.mean()
mean_b = group_b.mean()

lift = (mean_b - mean_a) / mean_a
print(f"Mean A (False): {mean_a:.4f}")
print(f"Mean B (True): {mean_b:.4f}")
print(f"Lift: {lift*100:.2f}%")


Mean A (False): 27.3968
Mean B (True): 27.4730
Lift: 0.28%


In [68]:
from scipy.stats import ttest_ind

t_stat, p_value = ttest_ind(group_a, group_b, equal_var=False)
print(f"T-statistic: {t_stat:.4f}, P-value: {p_value:.4f}")


T-statistic: -0.2468, P-value: 0.8051


In [69]:
df['host_identity_verified'] = df['host_identity_verified'].replace({'t': True, 'f': False, 'True': True, 'False': False})
df['host_identity_verified'] = df['host_identity_verified'].astype(bool)


In [70]:
df = df[df['host_identity_verified'].notnull()]


In [72]:
df_ab2 = df[df['number of reviews'].notnull()]


In [74]:
group_a = df_ab2[df_ab2['host_identity_verified'] == False]['number of reviews']
group_b = df_ab2[df_ab2['host_identity_verified'] == True]['number of reviews']


In [75]:
mean_a = group_a.mean()
mean_b = group_b.mean()
lift = (mean_b - mean_a) / mean_a

print(f"Mean A (Unverified): {mean_a:.4f}")
print(f"Mean B (Verified): {mean_b:.4f}")
print(f"Lift: {lift*100:.2f}%")


Mean A (Unverified): nan
Mean B (Verified): 27.4347
Lift: nan%


In [76]:
df['host_identity_verified'].value_counts()

host_identity_verified
True    102599
Name: count, dtype: int64

In [79]:
df_ab = df[df['number of reviews'].notnull()]
df_ab = df_ab[df_ab['room type'].isin(['Entire home/apt', 'Private room'])]

group_a = df_ab[df_ab['room type'] == 'Private room']['number of reviews']
group_b = df_ab[df_ab['room type'] == 'Entire home/apt']['number of reviews']

mean_a = group_a.mean()
mean_b = group_b.mean()
lift = (mean_b - mean_a) / mean_a * 100

from scipy.stats import ttest_ind
t_stat, p_value = ttest_ind(group_a, group_b, equal_var=False)

print(f"Mean A (Private room): {mean_a:.2f}")
print(f"Mean B (Entire home): {mean_b:.2f}")
print(f"Lift: {lift:.2f}%")
print(f"P-value: {p_value:.4f}")


Mean A (Private room): 27.74
Mean B (Entire home): 27.34
Lift: -1.47%
P-value: 0.1952


The difference is small (only 1.47%) and not statistically significant (p > 0.05).

So, room type doesn’t seem to have a strong impact on review counts in this dataset — at least not between these two categories.

In [81]:
df_ab = df[df['number of reviews'].notnull()]
df_ab = df_ab[df_ab['cancellation_policy'].isin(['flexible', 'strict'])]

group_a = df_ab[df_ab['cancellation_policy'] == 'strict']['number of reviews']
group_b = df_ab[df_ab['cancellation_policy'] == 'flexible']['number of reviews']

mean_a = group_a.mean()
mean_b = group_b.mean()
lift = (mean_b - mean_a) / mean_a * 100

from scipy.stats import ttest_ind
t_stat, p_value = ttest_ind(group_a, group_b, equal_var=False)

print(f"Mean A (Strict): {mean_a:.2f}")
print(f"Mean B (Flexible): {mean_b:.2f}")
print(f"Lift: {lift:.2f}%")
print(f"P-value: {p_value:.4f}")


Mean A (Strict): 27.29
Mean B (Flexible): 27.52
Lift: 0.83%
P-value: 0.5495


The lift is very small: listings with flexible policies have just 0.83% more reviews on average.

The p-value is 0.5495, meaning the difference is not statistically significant.

So once again, no strong evidence that cancellation policy alone impacts review volume.

In [82]:
# Remove dollar signs and commas, then convert to float
df['price'] = df['price'].replace('[\$,]', '', regex=True).astype(float)


In [83]:
# Filter valid rows
df_ab = df[df['price'].notnull() & df['number of reviews'].notnull()]

# Create A/B groups
group_a = df_ab[df_ab['price'] < 100]['number of reviews']
group_b = df_ab[df_ab['price'] >= 100]['number of reviews']

# Calculate mean and lift
mean_a = group_a.mean()
mean_b = group_b.mean()
lift = (mean_b - mean_a) / mean_a * 100

# Run t-test
from scipy.stats import ttest_ind
t_stat, p_value = ttest_ind(group_a, group_b, equal_var=False)

# Print results
print(f"Mean A (< $100): {mean_a:.2f}")
print(f"Mean B (>= $100): {mean_b:.2f}")
print(f"Lift: {lift:.2f}%")
print(f"P-value: {p_value:.4f}")


Mean A (< $100): 25.75
Mean B (>= $100): 27.51
Lift: 6.85%
P-value: 0.0204


Listings priced ≥ $100 receive 6.85% more reviews on average.

The p-value is below 0.05, so this difference is statistically significant.

This suggests that higher-priced listings may actually attract more guest activity, contrary to common assumptions that cheaper = more bookings.

In [84]:
df['service fee'] = df['service fee'].replace('[\$,]', '', regex=True).astype(float)


In [85]:
# Filter valid rows
df_ab = df[df['service fee'].notnull() & df['number of reviews'].notnull()]

# Find the median
median_fee = df_ab['service fee'].median()

# Create A and B groups
group_a = df_ab[df_ab['service fee'] < median_fee]['number of reviews']
group_b = df_ab[df_ab['service fee'] >= median_fee]['number of reviews']

# Calculate mean and lift
mean_a = group_a.mean()
mean_b = group_b.mean()
lift = (mean_b - mean_a) / mean_a * 100

# Run t-test
from scipy.stats import ttest_ind
t_stat, p_value = ttest_ind(group_a, group_b, equal_var=False)

# Print results
print(f"Median Service Fee: {median_fee:.2f}")
print(f"Mean A (< median): {mean_a:.2f}")
print(f"Mean B (>= median): {mean_b:.2f}")
print(f"Lift: {lift:.2f}%")
print(f"P-value: {p_value:.4f}")


Median Service Fee: 125.00
Mean A (< median): 27.40
Mean B (>= median): 27.49
Lift: 0.34%
P-value: 0.7632


**Interpretation:**

The difference in review count between lower and higher service fees is extremely small (+0.34%) and not statistically significant.

In short: service fee amount doesn’t seem to affect guest review behavior.



In [86]:
top2 = df['neighbourhood group'].value_counts().nlargest(2).index.tolist()
print(top2)


['Manhattan', 'Brooklyn']


In [87]:
# Filter valid rows and restrict to top 2 neighbourhoods
df_ab = df[df['neighbourhood group'].isin(top2) & df['number of reviews'].notnull()]

# Create groups
group_a = df_ab[df_ab['neighbourhood group'] == top2[0]]['number of reviews']
group_b = df_ab[df_ab['neighbourhood group'] == top2[1]]['number of reviews']

# Calculate means and lift
mean_a = group_a.mean()
mean_b = group_b.mean()
lift = (mean_b - mean_a) / mean_a * 100

# Run t-test
from scipy.stats import ttest_ind
t_stat, p_value = ttest_ind(group_a, group_b, equal_var=False)

# Print results
print(f"Comparing: {top2[0]} (A) vs. {top2[1]} (B)")
print(f"Mean A ({top2[0]}): {mean_a:.2f}")
print(f"Mean B ({top2[1]}): {mean_b:.2f}")
print(f"Lift: {lift:.2f}%")
print(f"P-value: {p_value:.4f}")


Comparing: Manhattan (A) vs. Brooklyn (B)
Mean A (Manhattan): 24.08
Mean B (Brooklyn): 28.46
Lift: 18.20%
P-value: 0.0000


**Interpretation:**
Listings in Brooklyn receive 18% more reviews on average than those in Manhattan.

The p-value is effectively zero, meaning this result is statistically very significant.

This suggests that Brooklyn listings may be more popular, affordable, or receive more consistent guest engagement — definitely something Airbnb might want to dig into further.